### PR TITLE
[DC] IAM Connector

### PR DESCRIPTION
**Pull request by alldoami**
Tested with python:
```
import src.connectors.aws_inventory as aws_inventory
options = {"connection_type": "IAM", "aws_access_key": "xxxx", "aws_secret_key": "xxxx", "accounts_connection_name": "AWS_ACCOUNTS_DEFAULT_CONNECTION"}
print([x for x in aws_inventory.ingest("AWS_ASSET_INV_IAM_DEFAULT_CONNECTION", options)])
```

### CODE REVIEW

File: src/connectors/aws_inventory.py [hunk 1]
```diff
@@ -327,6 +373,38 @@ def elb_dispatch(landing_table, aws_access_key='', aws_secret_key='', accounts=N
...
+    users = get_iam_users(
+        aws_access_key=aws_access_key,
+        aws_secret_key=aws_secret_key,
+        session=session,
+        account=account
+    )
+
+    monitor_time = datetime.utcnow().isoformat()
+
+    print(db.derive_insert_select(LANDING_TABLES_COLUMNS))
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]: rm these print statements

File: src/connectors/aws_inventory.py [hunk 2]
```diff
@@ -327,6 +373,38 @@ def elb_dispatch(landing_table, aws_access_key='', aws_secret_key='', accounts=N
...
+            monitor_time,
+            row['Path'],
+            row['UserName'],
+            row['UserId'],
+            row.get('Arn'),
+            row['CreateDate'],
+            row.get('PasswordLastUsed'))
+            for row in users
+        ],
+        select=db.derive_insert_select(LANDING_TABLES_COLUMNS),
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]: `LANDING_TABLE_COLUMNS` references all the landing tables for all types of ingestion. I think you want `LANDING_TABLES_COLUMNS['IAM']`

File: src/connectors/aws_inventory.py [hunk 3]
```diff
@@ -405,6 +483,33 @@ def ingest_elb(landing_table, aws_access_key=None, aws_secret_key=None, session=
     )
     return len(elbs)
 
+def get_iam_users(aws_access_key=None, aws_secret_key=None, session=None, account=None):
+    client = boto3.client(
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]: seems like this client is unused, you can remove it

File: src/connectors/aws_inventory.py [hunk 4]
```diff
@@ -145,6 +146,17 @@
         ('region_name', 'STRING(16)'),
         ('scheme', 'STRING(30)'),
         ('vpc_id', 'STRING(30)')
+    ],
+    # IAM Users
+    'IAM': [
+        ('raw', 'VARCHAR'),
+        ('ingested_at', 'TIMESTAMP_LTZ'),
+        ('Path', 'VARCHAR'),
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]: PATH

File: src/connectors/aws_inventory.py [hunk 5]
```diff
@@ -145,6 +146,17 @@
...
+        ('UserName', 'VARCHAR'),
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]: USER_NAME

File: src/connectors/aws_inventory.py [hunk 6]
```diff
@@ -145,6 +146,17 @@
...
+        ('UserId', 'VARCHAR'),
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]: USER_ID

File: src/connectors/aws_inventory.py [hunk 7]
```diff
@@ -145,6 +146,17 @@
...
+        ('Arn', 'VARCHAR'),
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]:
```suggestion
        ('arn', 'VARCHAR'),
```

File: src/connectors/aws_inventory.py [hunk 8]
```diff
@@ -145,6 +146,17 @@
...
+        ('CreateDate', 'TIMESTAMP_LTZ'),
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]:
```suggestion
        ('create_date', 'TIMESTAMP_LTZ'),
```

File: src/connectors/aws_inventory.py [hunk 9]
```diff
@@ -145,6 +146,17 @@
...
+        ('PasswordLastUsed', 'TIMESTAMP_LTZ')
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]:
```suggestion
        ('password_last_used', 'TIMESTAMP_LTZ')
```

File: src/connectors/aws_inventory.py (hunk 4 above)
Comment (edulop91) [reply src/connectors/aws_inventory.py]: all of these for convention and consistency

File: src/connectors/aws_inventory.py (hunk 3 above)
Comment (alldoami) [reply src/connectors/aws_inventory.py]: We still need to initialize it before the if statement right?

**Comment by greg-snowflake** [comment]
This looks good to me; can you add a note describing how you tested it?

**Comment by edulop91** [comment]
python:
```
import src.connectors.aws_inventory as aws_inventory
options = {"connection_type": "IAM", "aws_access_key": "xxxx", "aws_secret_key": "xxxx", "accounts_connection_name": "AWS_ACCOUNTS_DEFAULT_CONNECTION"}
print([x for x in aws_inventory.ingest("AWS_ASSET_INV_IAM_DEFAULT_CONNECTION", options)])
```

**Comment by andrey-snowflake** [comment]
lgtm, although I think this is doubling up with our AWS Config ingestion, which contains IAM changes, as well. there's also a config snapshot API call we could use, but with other APIs this has been the most thorough approach.

since there isn't anything super urgent on the release line right now, I think we can include this in v1.8.7 and push back the release a week (leaving next week for testing)?

**Comment by omer-snowflake** [comment]
@andrey-snowflake actually this is information that is not brought back by Config. As I was working on CIS AWS Benchmark last week I saw we were missing important details like MFA and last login for users. This was a story in our upcoming sprint so we'll definitely want this one in the next release (pushing it back a week sounds fine to allow for testing). Thank you @alldoami and @edulop91 for putting this one together! I'll share the CIS compliance rules when we have them working with this data.

File: src/connectors/aws_inventory.py [hunk 10]
```diff
@@ -405,6 +483,30 @@ def ingest_elb(landing_table, aws_access_key=None, aws_secret_key=None, session=
...
+    page_iterator = paginator.paginate()
+    results = [
+        user
+        for page in page_iterator
+        for user in page['Users']
+    ]
+
+    for user in results:
+        if account:
+            results['Account'] = account
```
Comment (edulop91) [inline src/connectors/aws_inventory.py]:
```suggestion
            user['Account'] = account
```

**Pull request merged**